app/api/entities/v2_support.ts

```diff
@@ -1,67 +1,188 @@
-}) => {
-  const transactionManager = DefaultTransactionManager();
-  if (await DefaultSettingsDataSource(transactionManager).readNewRelationshipsAllowed()) {
+const denormalizeAfterEntityAction = async (
+  {
+    sharedId,
+    language,
+  }: {
+    sharedId: string;
+    language: string;
+  },
+  method: 'Creating' | 'Updating'
+) => {
+  if (await newRelationshipsEnabled()) {
+    const transactionManager = DefaultTransactionManager();
     const denormalizationService = await DenormalizationService(transactionManager);
-    await denormalizationService.denormalizeAfterCreatingEntities([sharedId], language);
+    await denormalizationService[`denormalizeAfter${method}Entities`]([sharedId], language);
     await transactionManager.executeOnCommitHandlers(undefined);
   }
 };
+const denormalizeAfterEntityCreation = async (data: { sharedId: string; language: string }) =>
+  denormalizeAfterEntityAction(data, 'Creating');
 
-const denormalizeAfterEntityUpdate = async ({
-  sharedId,
-  language,
-}: {
-  sharedId: string;
-  language: string;
-}) => {
-  const transactionManager = DefaultTransactionManager();
-  if (await DefaultSettingsDataSource(transactionManager).readNewRelationshipsAllowed()) {
-    const denormalizationService = await DenormalizationService(transactionManager);
-    await denormalizationService.denormalizeAfterUpdatingEntities([sharedId], language);
-    await transactionManager.executeOnCommitHandlers(undefined);
-  }
+const denormalizeAfterEntityUpdate = async (data: { sharedId: string; language: string }) =>
+  denormalizeAfterEntityAction(data, 'Updating');
+
+const calculateDiff = (currentDoc: EntitySchema, toSave: EntitySchema, name: string) => {
+  const newValues = new Set<string>();
+  const deletedValues = new Set<string>();
+
+  toSave.metadata?.[name]?.forEach(item => {
+    newValues.add(item.value as string);
+  });
+
+  currentDoc.metadata?.[name]?.forEach(item => {
+    const value = item.value as string;
+    if (!newValues.has(value)) {
+      deletedValues.add(value);
+    }
+    newValues.delete(value);
+  });
+
+  return { newValues: [...newValues], deletedValues: [...deletedValues] };
+};
+
+interface RelationshipDefinition {
+  type: string;
+  to: string;
+  from: string;
+}
+
+interface DefinitionsToUpdate {
+  newRelationships: RelationshipDefinition[];
+  removedRelationships: RelationshipDefinition[];
+}
+
+const determineRelationships = async (
+  entitiesDataSource: EntitiesDataSource,
+  values: string[],
+  entity: EntitySchema,
+  query: MatchQueryNode
+) => {
+  const relationships: RelationshipDefinition[] = [];
+  await entitiesDataSource.getByIds(values, entity.language).forEach(async targetEntity => {
+    relationships.push(query.determineRelationship(targetEntity));
+  });
+  return relationships;
 };
 
 const ignoreNewRelationshipsMetadata = async (
   currentDoc: EntitySchema,
   toSave: EntitySchema,
   template: TemplateSchema
+): Promise<DefinitionsToUpdate> => {
```

Comment:
I don't fully understand the concern. But let me elaborate.
The `targetTemplates` property is only a calculated property, from the query, useful for the clients. The backend doesn't need the value because it doesn't even need to know what templates it would use, because it would just try to determine which relationship it needs to create/delete for the updated value. If the entity provided to the field is not of one of the `targetTemplates` it will throw an exception. In that line, I think what is missing is a validation on the rest api side, that could check that the provided entities are of one of those templates to prevent the exception on execution. Is that what you are refering to?
If that's the case, I agree and I can work on that.

Reply:
As discussed, now `targetTemplates` is used by the backend to perform validations on entity update/creation.